### DISCUSSION TITLE
How to set openssl as the default crypto backend?

### DISCUSSION THREAD

**Discussion by alireza11048** [strongswan/strongswan#512] [Q&A] [answered]
Hi,

I am trying to use OpenSSL with a specific engine as the crypto backend in the strongswan. successfully I've compiled the stonrgswan with the enabled OpenSSL plugin flag. also, I have configured the /etc/strongswan.d/charon/openssl.conf to be enabled and use my specific crypto engine. 
I should note that I can successfully establish an IPsec tunnel with the ipsec tool. the problem is that I think that strongswan doesn't use the openssl to do cryptographic primitives because my crypto engine does not report any activity, though I can successfully establish and use IPSec tunnel.

Is there any configuration which I should set in order to force strongswan to use the openssl plugin for the cryptographic primitives? 

thanks 

P.S, this is the ipsec listalgs result
List of registered IKE algorithms:

```
  encryption: AES_CBC[aes] AES_ECB[aes] 3DES_CBC[des] DES_CBC[des] DES_ECB[des] RC2_CBC[rc2] CAMELLIA_CBC[openssl]
              CAST_CBC[openssl] BLOWFISH_CBC[openssl] IDEA_CBC[openssl] NULL[openssl]
  integrity:  HMAC_MD5_96[openssl] HMAC_MD5_128[openssl] HMAC_SHA1_96[openssl] HMAC_SHA1_128[openssl]
              HMAC_SHA1_160[openssl] HMAC_SHA2_256_128[openssl] HMAC_SHA2_256_256[openssl] HMAC_SHA2_384_192[openssl]
              HMAC_SHA2_384_384[openssl] HMAC_SHA2_512_256[openssl] HMAC_SHA2_512_512[openssl] CAMELLIA_XCBC_96[xcbc]
              AES_XCBC_96[xcbc] AES_CMAC_96[cmac]
  aead:       AES_GCM_16[openssl] AES_GCM_12[openssl] AES_GCM_8[openssl]
  hasher:     HASH_SHA1[sha1] HASH_SHA2_224[sha2] HASH_SHA2_256[sha2] HASH_SHA2_384[sha2] HASH_SHA2_512[sha2]
              HASH_MD5[md5] HASH_MD4[openssl] HASH_IDENTITY[curve25519]
  prf:        PRF_KEYED_SHA1[sha1] PRF_HMAC_MD5[openssl] PRF_HMAC_SHA1[openssl] PRF_HMAC_SHA2_256[openssl]
              PRF_HMAC_SHA2_384[openssl] PRF_HMAC_SHA2_512[openssl] PRF_FIPS_SHA1_160[fips-prf] PRF_AES128_XCBC[xcbc]
              PRF_CAMELLIA128_XCBC[xcbc] PRF_AES128_CMAC[cmac]
  xof:        XOF_MGF1_SHA1[mgf1] XOF_MGF1_SHA224[mgf1] XOF_MGF1_SHA256[mgf1] XOF_MGF1_SHA384[mgf1]
              XOF_MGF1_SHA512[mgf1]
  drbg:       DRBG_CTR_AES128[drbg] DRBG_CTR_AES192[drbg] DRBG_CTR_AES256[drbg] DRBG_HMAC_SHA1[drbg]
              DRBG_HMAC_SHA256[drbg] DRBG_HMAC_SHA384[drbg] DRBG_HMAC_SHA512[drbg]
  dh-group:   ECP_256[openssl] ECP_384[openssl] ECP_521[openssl] ECP_224[openssl] ECP_192[openssl] ECP_256_BP[openssl]
              ECP_384_BP[openssl] ECP_512_BP[openssl] ECP_224_BP[openssl] MODP_3072[openssl] MODP_4096[openssl]
              MODP_6144[openssl] MODP_8192[openssl] MODP_2048[openssl] MODP_2048_224[openssl] MODP_2048_256[openssl]
              MODP_1536[openssl] MODP_1024[openssl] MODP_1024_160[openssl] MODP_768[openssl] MODP_CUSTOM[openssl]
              CURVE_25519[curve25519]
  random-gen: RNG_WEAK[openssl] RNG_STRONG[random] RNG_TRUE[random]
  nonce-gen:  [nonce]
```

And this is my ipsec.conf file:
```
config setup
        charondebug="all"
        uniqueids=yes
conn prodgateway-to-devgateway
        type=tunnel
        auto=start
        keyexchange=ikev2
        authby=secret
        left=192.168.99.17
        leftsubnet=192.168.99.17/24
        right=192.168.99.11
        rightsubnet=192.168.99.11/24
        ike=aes256-sha1-modp1024!
        esp=aes256-sha1!
        aggressive=no
        keyingtries=%forever
        ikelifetime=28800s
        lifetime=3600s
        dpddelay=30s
        dpdtimeout=120s
        dpdaction=restart
```

**Comment by tobiasbrunner** [comment, marked as answer]
What algorithm did you want to handle via _openssl_ plugin? If it's AES, you can clearly see that it's handled by the _aes_ plugin. So either prevent that from getting loaded or increase the priority of the _openssl_ plugin (see [PluginLoad](https://wiki.strongswan.org/projects/strongswan/wiki/PluginLoad)).

**Reply by alireza11048** [reply]
Increasing priority of the OpenSSL plugin worked for me, thank you

**Reply by alireza11048** [reply]
Just another question, 

Now I can see activity in my OpenSSL engine when I run the "ipsec restart" command (my engine reports 1-2 packet encryption/decryption), but when transferring a bulk of data through the tunnel, again I can't see activity on my engine! it seems that strongswan uses my engine just for initiating the connection, not for encryption the data which is transmitted over the tunnel.

could you please tell me that is there any specific configuration that should be done to use the engine to encrypting the data which is transmitted over the tunnel?

thanks

**Reply by tobiasbrunner** [reply]
IPsec traffic is handled in the OS kernel (only via [kernel-libipsec](https://wiki.strongswan.org/projects/strongswan/wiki/kernel-libipsec) is traffic handled by strongSwan in the userland - make sure to read that page thoroughly before considering using it). Please read [the introduction to strongSwan](https://wiki.strongswan.org/projects/strongswan/wiki/IntroductionTostrongSwan) for details.

**Reply by alireza11048** [reply]
Thank you so much,

As I understand it, by default the actual encryption/decryption of the IPsec traffic is handled in the Linux Kernel, and by enabling the kernel-libipsec encryption/decryption configuration would be handled by the Strongswan and OpenSSL plugin.

Maybe I shouldn't question in this forum about this, but could you please give me a clue to configure Linux IPSec stack behavior to use hardware accelerators for encryption/decryption of the traffic? (again, excuse me for irrelevant question)

thank you

**Reply by Thermi** [reply]
The ciphers are instanciated using the kernel crypto subsystem. It follows the same rules for implementation selection as other subsystems using it.

**Reply by alireza11048** [reply]
Thanks

**Reply by alireza11048** [reply]
Just another question, 

If the real implementation of the encryption/decryption is done in the kernel of the linux, what is purpose of different plugins in the strongswan, like AES or OpenSSL plugins?

thanks

**Reply by tobiasbrunner** [reply]
> If the real implementation of the encryption/decryption is done in the kernel of the linux, what is purpose of different plugins in the strongswan, like AES or OpenSSL plugins?

IKE (see the link I posted above for details).
